### PR TITLE
Implement DeepDDS

### LINKED ISSUE #19: Add the DeepDDS model
- Please read the paper first. It is [here](https://arxiv.org/abs/2107.02467).
- There is also code-release with the paper [here](https://github.com/Sinwang404/DeepDDs).
- After that read the contributing guidelines.
- If there is an existing open-source version of the model please take a look.
- ChemicalX is built on top of [PyTorch 1.10.](https://pytorch.org/) and [torchdrug](https://torchdrug.ai/).
- The library heavily builds on top on [torchdrug](https://torchdrug.ai/) and molecules in batches are PackedGraphs.
- There is already a model class under `./chemicalx/models/`
- Context features, drug level features, and labels are all FloatTensors.
- Look at the examples and tests under `./examples/` and `./tests/`.
- Add auxiliary layers as you see fit - please document these, add tests, and add these layers to the main readme.md if needed.
- Add typing to the initialization and forward pass.
- Non-data-dependent hyper-ammeters should have default values. 
- Please add tests under `./tests/` and make sure that your model/layer is tested with real data.
- Write an example under `./examples/`. What is the AUC on the test set? Is it reasonable?

### PR DESCRIPTION
**Pull request by kkaris**
Closes #19

Adds the DeepDDS model implementation
 
- [ ] Code passes all tests
- [x] Unit tests provided for these changes
- [x] Documentation and docstrings added for these changes

## Changes 

* Add DeepDDS
* Add new file, `deepdds_examples.py` containing an example

### CODE REVIEW

**Comment by kkaris** [comment]
This branch obviously need more work, but in for now I have this question:
- There are two version of this model, one using GAT and one using GCN for the drug feature embedding. Is the idea to implement both of them in the same module and also duplicate examples, tests etc? Or do we pick one of the implementations?

Another comment: it would be nice to have the shapes/sized of the input data either available somewhere as importable constants or automatically set by the pipeline or some wrapper to the pipeline.

**Comment by benedekrozemberczki** [comment]
What could be done is a parameter which deploys gat/gcn in the main model. That way you can test the model with various parameter settings in the tests and a single example enough. What do you think?

File: chemicalx/models/deepdds.py [hunk 1]
```diff
@@ -1,14 +1,106 @@
...
+    ):
+        """Instantiate the DeepDDS model.
+
+        :param context_feature_size:
+        :param context_output_size:
+        :param in_channels:
+        :param dropout:
+        """
+        super(DeepDDS, self).__init__()
+        self.cell_mlp = MLP(input_dim=context_feature_size, hidden_dims=[2048, 512, context_output_size])
```
Comment (benedekrozemberczki) [inline chemicalx/models/deepdds.py:59]: Should be parametrized.

File: chemicalx/models/deepdds.py [hunk 2]
```diff
@@ -1,14 +1,106 @@
...
+        self.conv_left = GraphConvolutionalNetwork(in_channels, [in_channels, in_channels * 2, in_channels * 4])
+        self.conv_right = GraphConvolutionalNetwork(in_channels, [in_channels, in_channels * 2, in_channels * 4])
+        self.mlp_left = MLP(
+            input_dim=in_channels * 4, hidden_dims=[in_channels * 2, context_output_size], dropout=dropout
+        )
+        self.mlp_right = MLP(input_dim=in_channels * 4, hidden_dims=[in_channels * 2, context_output_size])
+        self.mlp_final = MLP(context_output_size * 3, [1024, 512, 128, 1], dropout=dropout)
```
Comment (benedekrozemberczki) [inline chemicalx/models/deepdds.py:66]: Should be parametrized.

File: chemicalx/models/deepdds.py [hunk 3]
```diff
@@ -1,14 +1,106 @@
...
+
+    def unpack(self, batch: DrugPairBatch):
+        """Return the context features, left drug features and right drug features."""
+        return batch.context_features, batch.drug_molecules_left, batch.drug_molecules_right
+
+    def forward(
```
Comment (benedekrozemberczki) [inline chemicalx/models/deepdds.py:72]: Looks good.

**Review by benedekrozemberczki — COMMENTED** [review]
Looks good! Thank you for your contribution. The molecular encoders could be parametrically deployed and that way you could switch between GCN and GAT.

File: chemicalx/models/deepdds.py [hunk 4]
```diff
@@ -1,14 +1,106 @@
...
     """
+
+    # todo: implement the GAT version of the model as well
+
+    def __init__(
+        self,
+        *,
+        context_feature_size: int,
+        context_output_size: int,
+        in_channels: int,
```
Comment (cthoyt) [inline chemicalx/models/deepdds.py:48]: this should be 69

**Comment by benedekrozemberczki** [comment]
@kkaris thanks to @cthoyt now we have the default atomic feature count for the datasets.

File: examples/deepdds_examples.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+"""Example with DeepDDs."""
+
+from chemicalx import pipeline
+from chemicalx.models import DeepDDS
+
+
+def main():
+    """Train and evaluate the DeepDDs model."""
+    model = DeepDDS(
```
Comment (cthoyt) [inline examples/deepdds_examples.py:9]: follow pattern from other examples for pre-loading dataset and instantiating the model based on its properties

**Comment by cthoyt** [comment]
I think the reason this is broken is the PackedGraph object has a shape of `torch.Size([149868, 149868, 4])` instead of being a list of `batch_size` graphs. I think the solution is to apply a MaxReadout() at the end (which I think is the same as max pooling, but not sure)

**Comment by benedekrozemberczki** [comment]
Max readout pools the vectors and creates something that is batch_size X dim. It is a scattering transform - a way to think about it is a group by aggregate where nodes are grouped together by the source graph. 

https://github.com/rusty1s/pytorch_scatter 

The figure here is pretty great.

File: chemicalx/models/deepdds.py [hunk 6]
```diff
@@ -1,18 +1,184 @@
-"""An implementation of the DeepDDS model."""
+"""An implementation of the DeepDDS model.
 
-from .base import UnimplementedModel
```
Comment (benedekrozemberczki) [inline chemicalx/models/deepdds.py:3]: This is beautifully done.

File: chemicalx/models/deepdds.py [hunk 7]
```diff
@@ -1,18 +1,184 @@
...
+from chemicalx.data import DrugPairBatch
+from chemicalx.models import Model
 
 __all__ = [
     "DeepDDS",
 ]
 
 
-class DeepDDS(UnimplementedModel):
+class DeepDDS(Model):
```
Comment (benedekrozemberczki) [inline chemicalx/models/deepdds.py:42]: Looks good to mention both.

File: chemicalx/models/deepdds.py [hunk 8]
```diff
@@ -1,18 +1,184 @@
...
+        drug_channels: int = TORCHDRUG_NODE_FEATURES,
+        drug_gcn_hidden_dims: Optional[List[int]] = None,
+        drug_mlp_hidden_dims: Optional[List[int]] = None,
+        context_output_size: int = 128,
+        fc_hidden_dims: Optional[List[int]] = None,
+        dropout: float = 0.2,  # Rate used in paper
+    ):
+        """Instantiate the DeepDDS model.
+
+        :param context_feature_size:
```
Comment (benedekrozemberczki) [inline chemicalx/models/deepdds.py:70]: This is very detailed.

File: examples/deepdds_examples.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+from chemicalx.data import DrugCombDB
+from chemicalx.models import DeepDDS
+
+
+def main():
+    """Train and evaluate the DeepDDs model."""
+    dataset = DrugCombDB()
+    model = DeepDDS(
+        context_feature_size=dataset.context_channels,
+    )
```
Comment (benedekrozemberczki) [inline examples/deepdds_examples.py:13]: Looks good.

**Review by benedekrozemberczki — APPROVED** [review]
LGTM. That failure of the CI on the default values is weird.

**Comment by cthoyt** [comment]
The failure is because there are only a subset of special parameters (i.e., `drug_channels`, `context_channels`) that are allowed to not have a default.

**Comment by benedekrozemberczki** [comment]
@cthoyt Can I merge?
